### ISSUE TITLE
Enable flexible definition of asset name schemes from user

### ISSUE THREAD

**Issue by TobiasKadelka** [psyinfra/onyo#101]
I want to find a way to specify asset name schemes in onyo in a way that has at least be able to do:
- be used by `onyo new` to read a new asset file name
- `onyo set`, because it should not just be able to set contents, but also values for the name. Currently, there is no way to do this flexibly, except when assuming the name scheme will always be `<type>_<make>_<model>.<serial>` with the same reserved characters.
- `onyo mv --rename`, because currently this does not check names at all.
- `onyo fsck` should also check the name schemes
- define reserved characters for each asset template/type and for each field separately
- be compatible with future implementation of faux serials

------------

My thoughts so far were about adding a file `.onyo/validation/names.yaml` which could be easily used by `onyo new` and would be structured very similarly to the existing validation.yaml file.

```
default:
- name: <type>_<make>_<model>.<serial>
  - type:
    - reserved: _, .
  - make:
    - reserved: _, .
  - model:
    - reserved: _, .
<template name, e.g. laptop>:
  - name: <name_scheme>
```

But the problem is, I do not see how to use it with mv, fsck, set etc. All the ideas I had so far are based on "saving additional information" for each file:
- either in the asset
- in the .anchor
-  or even in an additional file

But this would directly break when changing a template name and force the user to update the names.yaml file everytime a new template would be added.

One could also track the templates and validation files with git and try to follow the information trough commits, but when then  the history would not be complete (manual rebase?), this might break the functionality again.

So long story short, I currently am still struggling to find a good mechanism to specify asset file names, that goes farther than asset creation with `onyo new`.

**Comment by TobiasKadelka** [comment]
Edit: I added to the above list that it should be compatible with "faux serials" that are mentioned in the readme.
This is not implemented right now, but whatever solution for asset names and their validation is found, should be compatible with the faux serials.

**Comment by TobiasKadelka** [comment]
This comment from #15 seems now more relevant in this context:

When running `onyo mv <path/filename> <path/new_filename>` without the rename flag, it mentions the flag as a forced solution. In the future, when onyo set is implemented, it should also mention `onyo set key=value` as a possibility.

**Comment by bpoldrack** [comment]
I'd like to provide a different angle on this.

What do we technically need from asset names?
- They are identifiers and need to be unique across the repo. This has no implications on what a name consists of.
- IF they contain pseudo-keys, we need to be able to extract their values from the name and set new values.

At the moment, I think that everything else is "made up" by having mentally settled on a combination of pseudo-keys to be a given. But that's not necessary. If a user comes up with unique identifiers and wants to track things like serial numbers, model names etc. in the YAML instead - that's perfectly fine and does not have impede any inventory functionality as far as I can see.

Therefore, I'd suggest to not enforce a particular scheme. Of course, by default we keep what we have, but what that really is about is pseudo keys, not naming scheme. Hence, make pseudo keys configurable by providing a regex that tells onyo how to extract a particular pseudo key (substring) from an asset name. If we can find the match, we can also replace the matching substring with a new value - no need to have an explicit definition for setting, I think. `.onyo/config` can then come with a pseudo key section that is preset to define our default ones, but can be changed to anything. This seems a lot easier to implement and define than any sort of scheme definition, while being very flexible.
For validation, we'd check whether names are unique and whether we can match the configured expressions.

WDYT, @TobiasKadelka, @nhjjr, @aghisla, @aqw?

**Comment by TobiasKadelka** [comment]
We had similar ideas in the past, but it was never fully sure how to do this, and some questions always remained, for example "is there just one or even many name schemes per repository?".

1. if just one scheme:
  - how to handle it when the scheme is changed later? Because `onyo set --rename` does change the keys separately, not the whole names.

2. if multiple ones:
  - how to adress them, when they are overlapping? (e.g. "type_make_model.serial" vs "day_week_month.year" ) I would guess we would always use the first scheme that fits?

3. how to verify/check the name-scheme itself? 
  - do we verify this file always when running onyo? (now that I think about this, we should maybe already do that, currently we just error when a tool specified under `onyo "history"` does error, but the name scheme is important for many more commands)
  - If this is broken, does onyo stop working? What if "path" is part of it, which is something that the output of `onyo get` already provides as a column. There might be similar bugs. 

In general I do agree and would love to have this implemented, I just have some open questions beforehand :)
But I think the general idea of flexible name schemes (e.g. user-definable pseudo-keys) was most of the time in our mind, and should be not that hard to implement.

The biggest TODO would probably be adding a lot of tests for this that verify that none of the existing commands have something about this hardcoded, and every command actually uses the name from the config correctly :)

**Comment by TobiasKadelka** [comment]
Also, this is then connected with https://github.com/psyinfra/onyo/issues/306 about more `.onyo/` support.

**Comment by bpoldrack** [comment]
@TobiasKadelka 
> "is there just one or even many name schemes per repository?"

If we go with the suggestion above, then there's no scheme. ;-)
There are pseudo key definitions and obviously you can't configure the same pseudo key several times.

**Comment by bpoldrack** [comment]
> how to adress them, when they are overlapping? (e.g. "type_make_model.serial" vs "day_week_month.year" ) I would guess we would always use the first scheme that fits?

Both and I think that's the user's problem.
Your example would mean to specify a pseudo key `type` and a pseudo key `day` with the same regex. We'd therefore use that regex whenever the respective key accessed. If a user sets it up that way - not onyo's problem, I think. (There could be valid cases for that)

**Comment by bpoldrack** [comment]
> do we verify this file always when running onyo? (now that I think about this, we should maybe already do that, currently we just error when a tool specified under onyo "history" does error, but the name scheme is important for many more commands)

I don't think there's anything to verify. It's a regex. We apply it. If that fails, we fail.

**Comment by nhjjr** [comment]
> Therefore, I'd suggest to not enforce a particular scheme. Of course, by default we keep what we have, but what that really is about is pseudo keys, not naming scheme.

I agree and propose (a possibly unpopular opinion) of taking this a step further. Scrap any internal reliance on pseudo-keys.

What do pseudo-keys currently do for us?
  1. Provide a (default) naming scheme for asset files
  2. Enable users to easily explore the inventory using a file browser

This leads to a set of problems for us:
  1. We don't want users to modify files through the file browser, so why do we make it seem as if that is encouraged?
  2. We have to provide workarounds for certain functions, e.g.:
    * `onyo mv` cannot rename files because it may break pseudo-keys
    * `onyo get` must search separately through file names _and_ file contents
  3. Workarounds will become more entrenched and thus difficult to modify (unless preempted by decoupling pseudo-key handling)
  4. Pseudo-keys are not straight-forward to new users who explore onyo through a file browser

This can be resolved by not giving a damn about file names, barring the extension which we'll use to ensure file uniqueness through a randomly generated serial. Two suggestions to deal with this:
  * When a new file is made, force the user to give that file a name
  * When a new file is made, give it a name based on required keys from the asset file
  * When a new file is made, give it a random name if no name is specified

Even if the file names become a randomized mess, users should be able to use `onyo get` and `onyo ls` to obtain identifiable assets. For example, an `onyo ls -k type make model` could return something like:

```
total 5 (1 folder, 4 assets)
/folder                
a89d-89bd-kk87.83hic    monitor  dell    SE2722HX  
wintermute.9lopi        laptop   apple   macbookpro
bens-laptop.3pop5       laptop   lenovo  thinkpad
laptop_hp_zbook.876ab   laptop   hp      zbook
```

This way we never rely on filename to be a reflection of any particular value, allowing it to literally be anything. This gets rid of pseudo-keys and all the associated problems, but comes at the cost of developing more options for commands like `onyo ls`. This could even be extended with naming helpers, for example a function that can rename all files in a folder (or the entire repository) using asset keys (e.g., `onyo rename --all -k type make model --sep "_"` to rename all assets in a repository using their values of type, make, and model with an _ as a separator).

Thoughts? Or did I go too far?

**Comment by aqw** [comment]
@nhjjr Controversial is good. I appreciate you throwing a firework into the middle of everything. It's good for stimulation. ;-)

Reading through this, it feels like it would set us on the path of implementing an RDBMS on-top of git. At that point, I think we should simply use sqlite and dump it to git for syncing.

I'm pretty attached to the obviousness (aka: intuitiveness) of using file names that are meaningful to a user. I /do/ agree that some of the lengths onyo goes to in order to keep that sane is a bit suboptimal at times. I think your criticisms (especially 2 and 3) are well founded.

I see onyo from a very unix-based view: everything is a file. Get to use all the unix tools on them. So I will keep coming back to that.

One point: the term "pseudo keys" (as it's currently understand) is a category that's larger than just the keys in the filenames. It also includes things like `path`  and possibly others (last commit time, etc).

Two repond to two points of yours:

> * onyo mv cannot rename files because it may break pseudo-keys

IMO, this is because the file contains key-data. And for that reason `set` is more appropriate.

> * onyo get must search separately through file names and file contents

That is annoying/unfortunate. This was decided so that the information is only ever in one place. Two sources don't have a tie breaker.

However, I could see taking a view that all keys must exist inside the file, and that the filename is merely an outward representation of a selection of keys that exist in the file. If there's a mismatch, the filename is considered wrong.

That way, the filename scheme could be updated/changed by the user, without there being any loss of data. Sure there's be a big commit changing all filenames. But no file contents would change. And it should simplify querying data.

Anyway. Those are my late night thoughts. :-)

**Comment by bpoldrack** [comment]
Verdict from roadmap meeting for 1.0:

- Make actual asset name scheme configurable by defining an f-string allowing to fill in any key from asset content like `"{key1}_{key2}.{key3}"`
- Support pseudo-key configuration like suggested above (regex for matching substring of filename), but read-only. Have reserved (not configurable) pseudo keys like `path`, `filename` and a couple of dates (last modified, created, etc.). TODO: Make a list of names+definitions for those.
